File: java/0005-longest-palindromic-substring.java
/**
 * LeetCode 0005: Longest Palindromic Substring
 *
 * Problem Statement:
 * Given a string s, return the longest palindromic substring in s.
 *
 * Example 1:
 * Input: s = "babad"
 * Output: "bab"
 * Explanation: "aba" is also a valid answer.
 *
 * Example 2:
 * Input: s = "cbbd"
 * Output: "bb"
 *
 * Constraints:
 * - 1 <= s.length <= 1000
 * - s consist of only digits and English letters.
 */

public class LongestPalindromicSubstring {

    // Q1: What's the brute force approach and its time complexity?
    // A1: Check all possible substrings for palindrome. Time: O(n^3), Space: O(1)
    public String longestPalindrome_bruteForce(String s) {
        int n = s.length();
        String longest = "";

        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                String substr = s.substring(i, j + 1);
                if (isPalindrome(substr) && substr.length() > longest.length()) {
                    longest = substr;
                }
            }
        }

        return longest;
    }

    private boolean isPalindrome(String s) {
        int left = 0, right = s.length() - 1;
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

    // Q2: Can we optimize by expanding around centers?
    // A2: Yes! Expand around each center. Time: O(n^2), Space: O(1)
    public String longestPalindrome_expandAroundCenter(String s) {
        if (s == null || s.length() < 1) {
            return "";
        }

        int start = 0;
        int maxLen = 0;

        for (int i = 0; i < s.length(); i++) {
            // Odd length palindromes (center is a single character)
            int len1 = expandAroundCenter(s, i, i);
            // Even length palindromes (center is between two characters)
            int len2 = expandAroundCenter(s, i, i + 1);

            int currentLen = Math.max(len1, len2);

            if (currentLen > maxLen) {
                maxLen = currentLen;
                start = i - (currentLen - 1) / 2;
            }
        }

        return s.substring(start, start + maxLen);
    }

    private int expandAroundCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        return right - left - 1;
    }

    // Q3: How does dynamic programming help solve this?
    // A3: Build table where dp[i][j] = true if s[i:j+1] is palindrome. Time: O(n^2), Space: O(n^2)
    public String longestPalindrome_dp(String s) {
        int n = s.length();
        if (n < 2) {
            return s;
        }

        boolean[][] dp = new boolean[n][n];

        // Every single character is a palindrome
        for (int i = 0; i < n; i++) {
            dp[i][i] = true;
        }

        int start = 0;
        int maxLen = 1;

        // Check for two-character palindromes
        for (int i = 0; i < n - 1; i++) {
            if (s.charAt(i) == s.charAt(i + 1)) {
                dp[i][i + 1] = true;
                start = i;
                maxLen = 2;
            }
        }

        // Check for palindromes of length 3 and more
        for (int length = 3; length <= n; length++) {
            for (int i = 0; i <= n - length; i++) {
                int j = i + length - 1;

                if (s.charAt(i) == s.charAt(j) && dp[i + 1][j - 1]) {
                    dp[i][j] = true;
                    start = i;
                    maxLen = length;
                }
            }
        }

        return s.substring(start, start + maxLen);
    }

    // Q4: What's Manacher's algorithm and how does it work?
    // A4: Linear time algorithm using symmetry. Time: O(n), Space: O(n)
    public String longestPalindrome_manacher(String s) {
        if (s == null || s.length() == 0) {
            return "";
        }

        // Transform string to handle even and odd length palindromes
        String T = preprocessString(s);
        int n = T.length();
        int[] P = new int[n];
        int C = 0; // Center
        int R = 0; // Right boundary

        for (int i = 1; i < n - 1; i++) {
            int mirror = 2 * C - i;

            if (i < R) {
                P[i] = Math.min(R - i, P[mirror]);
            }

            // Try to expand palindrome centered at i
            try {
                while (T.charAt(i + 1 + P[i]) == T.charAt(i - 1 - P[i])) {
                    P[i]++;
                }
            } catch (StringIndexOutOfBoundsException e) {
                // Do nothing
            }

            // If palindrome centered at i extends past R, adjust C and R
            if (i + P[i] > R) {
                C = i;
                R = i + P[i];
            }
        }

        // Find the maximum element in P
        int maxLen = 0;
        int centerIndex = 0;
        for (int i = 1; i < n - 1; i++) {
            if (P[i] > maxLen) {
                maxLen = P[i];
                centerIndex = i;
            }
        }

        int start = (centerIndex - maxLen) / 2;
        return s.substring(start, start + maxLen);
    }

    private String preprocessString(String s) {
        StringBuilder sb = new StringBuilder();
        sb.append('^');
        for (int i = 0; i < s.length(); i++) {
            sb.append('#').append(s.charAt(i));
        }
        sb.append('#').append('$');
        return sb.toString();
    }

    // Q5: How to count all palindromic substrings?
    // A5: Similar to expand around center but count all palindromes
    public int countPalindromicSubstrings(String s) {
        int total = 0;
        for (int i = 0; i < s.length(); i++) {
            total += countPalindromesAroundCenter(s, i, i);     // Odd length
            total += countPalindromesAroundCenter(s, i, i + 1); // Even length
        }
        return total;
    }

    private int countPalindromesAroundCenter(String s, int left, int right) {
        int count = 0;
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            count++;
            left--;
            right++;
        }
        return count;
    }

    // Q6: How do we handle edge cases?
    // A6: Empty strings, single character, entire string is palindrome
    public String longestPalindrome_withValidation(String s) {
        if (s == null || s.isEmpty()) {
            return "";
        }

        if (s.length() == 1) {
            return s;
        }

        // Check if entire string is palindrome
        if (isPalindrome(s)) {
            return s;
        }

        return longestPalindrome_expandAroundCenter(s);
    }

    // Q7: What's the space-time tradeoff analysis?
    /**
     * A7: Different approaches and their complexities:
     *
     * 1. Brute Force:
     *    - Time: O(n^3) - O(n^2) substrings, O(n) to check each
     *    - Space: O(1)
     *
     * 2. Dynamic Programming:
     *    - Time: O(n^2)
     *    - Space: O(n^2) - for DP table
     *
     * 3. Expand Around Center:
     *    - Time: O(n^2) - O(n) centers, O(n) expansion each
     *    - Space: O(1)
     *    - Best practical approach for most cases
     *
     * 4. Manacher's Algorithm:
     *    - Time: O(n)
     *    - Space: O(n)
     *    - Optimal complexity but more complex to implement
     */

    // Q8: Can we optimize the expand around center approach?
    // A8: We can track indices instead of creating substrings
    public String longestPalindrome_optimized(String s) {
        if (s == null || s.length() < 1) {
            return "";
        }

        int start = 0;
        int end = 0;

        for (int i = 0; i < s.length(); i++) {
            int len1 = expandAroundCenter(s, i, i);
            int len2 = expandAroundCenter(s, i, i + 1);
            int len = Math.max(len1, len2);

            if (len > end - start) {
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }

        return s.substring(start, end + 1);
    }

    // Q9: How to implement with early termination optimization?
    // A9: If remaining string can't exceed current max, stop early
    public String longestPalindrome_earlyTermination(String s) {
        if (s == null || s.length() < 1) {
            return "";
        }

        int start = 0;
        int maxLen = 1;

        for (int i = 0; i < s.length(); i++) {
            // Early termination: if remaining string can't beat current max
            if (s.length() - i <= maxLen / 2) {
                break;
            }

            int len1 = expandAroundCenter(s, i, i);
            int len2 = expandAroundCenter(s, i, i + 1);
            int currentLen = Math.max(len1, len2);

            if (currentLen > maxLen) {
                maxLen = currentLen;
                start = i - (currentLen - 1) / 2;
            }
        }

        return s.substring(start, start + maxLen);
    }

    // Q10: What are common variations and related problems?
    /**
     * A10: Common variations include:
     *
     * 1. Palindrome Permutation: Check if string can be rearranged to form palindrome
     * 2. Valid Palindrome: Check if string is palindrome (ignoring non-alphanumeric)
     * 3. Valid Palindrome II: Allow deleting at most one character
     * 4. Palindromic Substrings: Count all palindromic substrings
     * 5. Shortest Palindrome: Add minimum characters to make palindrome
     * 6. Palindrome Partitioning: Partition string so each substring is palindrome
     * 7. Longest Palindromic Subsequence: Similar but non-contiguous
     *
     * Key insights:
     * - Expand around center is most practical
     * - DP good for learning but uses more space
     * - Manacher's for optimal complexity when needed
     * - Understanding palindrome properties is crucial
     */

    // Main method for testing
    public static void main(String[] args) {
        LongestPalindromicSubstring solution = new LongestPalindromicSubstring();

        // Test cases
        String[][] testCases = {
            {"babad", "bab"},  // "aba" is also valid
            {"cbbd", "bb"},
            {"a", "a"},
            {"ac", "a"},       // or "c"
            {"racecar", "racecar"},
            {"noon", "noon"},
            {"aaaa", "aaaa"},
            {"abacabad", "abacaba"}
        };

        System.out.println("Testing Expand Around Center Solution (Recommended):");
        for (String[] testCase : testCases) {
            String input = testCase[0];
            String result = solution.longestPalindrome_expandAroundCenter(input);
            System.out.println("Input: s = \"" + input + "\"");
            System.out.println("Output: \"" + result + "\"");
            System.out.println("Length: " + result.length());
            System.out.println();
        }

        System.out.println("\nTesting Manacher's Algorithm (O(n) time):");
        for (int i = 0; i < Math.min(5, testCases.length); i++) {
            String input = testCases[i][0];
            String result = solution.longestPalindrome_manacher(input);
            System.out.println("Input: \"" + input + "\" -> Output: \"" + result + "\"");
        }

        System.out.println("\nCounting palindromic substrings:");
        String[] countTests = {"abc", "aaa", "noon"};
        for (String s : countTests) {
            int count = solution.countPalindromicSubstrings(s);
            System.out.println("Input: \"" + s + "\" has " + count + " palindromic substrings");
        }
    }
}

File: python/0005-longest-palindromic-substring.py
"""
LeetCode 0005: Longest Palindromic Substring

Problem Statement:
Given a string s, return the longest palindromic substring in s.

Example 1:
Input: s = "babad"
Output: "bab"
Explanation: "aba" is also a valid answer.

Example 2:
Input: s = "cbbd"
Output: "bb"

Constraints:
- 1 <= s.length <= 1000
- s consist of only digits and English letters.
"""

# Practice Questions and Answers

# Q1: What's the brute force approach and its time complexity?
# A1: Check all possible substrings for palindrome. Time: O(n^3), Space: O(1)
def longestPalindrome_brute_force(s):
    def is_palindrome(substr):
        return substr == substr[::-1]

    n = len(s)
    longest = ""

    for i in range(n):
        for j in range(i, n):
            substr = s[i:j+1]
            if is_palindrome(substr) and len(substr) > len(longest):
                longest = substr

    return longest

# Q2: Can we optimize by expanding around centers?
# A2: Yes! Expand around each center (single char and between chars). Time: O(n^2), Space: O(1)
def longestPalindrome_expand_around_center(s):
    def expand_around_center(left, right):
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return right - left - 1

    if not s:
        return ""

    start = 0
    max_len = 0

    for i in range(len(s)):
        # Odd length palindromes (center is a single character)
        len1 = expand_around_center(i, i)
        # Even length palindromes (center is between two characters)
        len2 = expand_around_center(i, i + 1)

        current_len = max(len1, len2)

        if current_len > max_len:
            max_len = current_len
            start = i - (current_len - 1) // 2

    return s[start:start + max_len]

# Q3: How does dynamic programming help solve this?
# A3: Build table where dp[i][j] = True if s[i:j+1] is palindrome. Time: O(n^2), Space: O(n^2)
def longestPalindrome_dp(s):
    n = len(s)
    if n < 2:
        return s

    # dp[i][j] = True if s[i:j+1] is a palindrome
    dp = [[False] * n for _ in range(n)]

    # Every single character is a palindrome
    for i in range(n):
        dp[i][i] = True

    start = 0
    max_len = 1

    # Check for two-character palindromes
    for i in range(n - 1):
        if s[i] == s[i + 1]:
            dp[i][i + 1] = True
            start = i
            max_len = 2

    # Check for palindromes of length 3 and more
    for length in range(3, n + 1):
        for i in range(n - length + 1):
            j = i + length - 1

            if s[i] == s[j] and dp[i + 1][j - 1]:
                dp[i][j] = True
                start = i
                max_len = length

    return s[start:start + max_len]

# Q4: What's Manacher's algorithm and how does it work?
# A4: Linear time algorithm using symmetry. Time: O(n), Space: O(n)
def longestPalindrome_manacher(s):
    """
    Manacher's algorithm for finding longest palindrome in O(n) time.
    Transform string to handle even/odd length palindromes uniformly.
    """
    # Transform string to handle even and odd length palindromes
    # "abc" -> "^#a#b#c#$"
    T = '#'.join('^{}$'.format(s))
    n = len(T)
    P = [0] * n  # P[i] = length of palindrome centered at i
    C = R = 0  # Center and right boundary of rightmost palindrome

    for i in range(1, n - 1):
        # Mirror of i with respect to C
        mirror = 2 * C - i

        if i < R:
            P[i] = min(R - i, P[mirror])

        # Try to expand palindrome centered at i
        try:
            while T[i + 1 + P[i]] == T[i - 1 - P[i]]:
                P[i] += 1
        except:
            pass

        # If palindrome centered at i extends past R, adjust C and R
        if i + P[i] > R:
            C, R = i, i + P[i]

    # Find the maximum element in P
    max_len, center_index = max((length, i) for i, length in enumerate(P))

    # Extract the longest palindrome from original string
    start = (center_index - max_len) // 2
    return s[start:start + max_len]

# Q5: How to find all palindromic substrings (not just the longest)?
# A5: Similar to expand around center but collect all palindromes
def count_palindromic_substrings(s):
    def expand_around_center(left, right):
        count = 0
        while left >= 0 and right < len(s) and s[left] == s[right]:
            count += 1
            left -= 1
            right += 1
        return count

    total = 0
    for i in range(len(s)):
        total += expand_around_center(i, i)  # Odd length
        total += expand_around_center(i, i + 1)  # Even length

    return total

# Q6: What if we want to return all longest palindromes (in case of ties)?
# A6: Track all palindromes with maximum length
def longestPalindrome_all_max(s):
    def expand_around_center(left, right):
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return left + 1, right - 1

    if not s:
        return []

    max_len = 0
    results = []

    for i in range(len(s)):
        # Odd length
        l1, r1 = expand_around_center(i, i)
        if r1 - l1 + 1 > max_len:
            max_len = r1 - l1 + 1
            results = [s[l1:r1+1]]
        elif r1 - l1 + 1 == max_len and s[l1:r1+1] not in results:
            results.append(s[l1:r1+1])

        # Even length
        l2, r2 = expand_around_center(i, i + 1)
        if r2 >= l2:
            if r2 - l2 + 1 > max_len:
                max_len = r2 - l2 + 1
                results = [s[l2:r2+1]]
            elif r2 - l2 + 1 == max_len and s[l2:r2+1] not in results:
                results.append(s[l2:r2+1])

    return results

# Q7: How do we handle edge cases?
# A7: Empty strings, single character, entire string is palindrome
def longestPalindrome_with_validation(s):
    if not s:
        return ""

    if len(s) == 1:
        return s

    # Check if entire string is palindrome
    if s == s[::-1]:
        return s

    # Use expand around center approach
    return longestPalindrome_expand_around_center(s)

# Q8: What's the space-time tradeoff analysis?
"""
A8: Different approaches and their complexities:

1. Brute Force:
   - Time: O(n^3) - O(n^2) substrings, O(n) to check each
   - Space: O(1)

2. Dynamic Programming:
   - Time: O(n^2)
   - Space: O(n^2) - for DP table

3. Expand Around Center:
   - Time: O(n^2) - O(n) centers, O(n) expansion each
   - Space: O(1)
   - Best practical approach for most cases

4. Manacher's Algorithm:
   - Time: O(n)
   - Space: O(n)
   - Optimal complexity but more complex to implement
"""

# Q9: How can we optimize space in the DP approach?
# A9: We only need the previous row, can reduce to O(n) space
def longestPalindrome_dp_space_optimized(s):
    n = len(s)
    if n < 2:
        return s

    # We only need to track if dp[i+1][j-1] was True
    # Can use a 1D array and update carefully
    prev = [False] * n
    curr = [False] * n

    start = 0
    max_len = 1

    # Single characters
    for i in range(n):
        curr[i] = True

    # Two characters
    for i in range(n - 1):
        if s[i] == s[i + 1]:
            start = i
            max_len = 2
            # Mark in a way we can track

    # For longer substrings, we need full 2D array
    # Space optimization is limited here
    return longestPalindrome_dp(s)

# Q10: What are common variations and related problems?
"""
A10: Common variations include:

1. Palindrome Permutation: Check if string can be rearranged to form palindrome
2. Valid Palindrome: Check if string is palindrome (ignoring non-alphanumeric)
3. Valid Palindrome II: Allow deleting at most one character
4. Palindromic Substrings: Count all palindromic substrings
5. Shortest Palindrome: Add minimum characters to make palindrome
6. Palindrome Partitioning: Partition string so each substring is palindrome
7. Longest Palindromic Subsequence: Similar but non-contiguous

Key insights:
- Expand around center is most practical
- DP good for learning but uses more space
- Manacher's for optimal complexity when needed
- Understanding palindrome properties is crucial
"""

if __name__ == "__main__":
    # Test cases
    test_cases = [
        ("babad", ["bab", "aba"]),  # Multiple valid answers
        ("cbbd", ["bb"]),
        ("a", ["a"]),
        ("ac", ["a", "c"]),
        ("racecar", ["racecar"]),
        ("noon", ["noon"]),
        ("abcdefg", None),  # Any single character
        ("aaaa", ["aaaa"]),
        ("abacabad", ["abacaba"]),
    ]

    print("Testing Expand Around Center Solution (Recommended):")
    for s, expected in test_cases:
        result = longestPalindrome_expand_around_center(s)
        print(f"Input: s = '{s}'")
        print(f"Output: '{result}'")
        if expected:
            is_valid = result in expected or len(result) == max(len(e) for e in expected)
            print(f"Valid: {is_valid} (Expected one of {expected})")
        else:
            print(f"Length: {len(result)}")
        print()

    print("\nTesting Manacher's Algorithm (O(n) time):")
    for s, _ in test_cases[:5]:
        result = longestPalindrome_manacher(s)
        print(f"Input: '{s}' -> Output: '{result}'")

    print("\nCounting palindromic substrings:")
    test_count = ["abc", "aaa", "noon"]
    for s in test_count:
        count = count_palindromic_substrings(s)
        print(f"Input: '{s}' has {count} palindromic substrings")
